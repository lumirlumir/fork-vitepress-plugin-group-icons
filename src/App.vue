<script setup lang="ts">
import GroupIcon from './components/GroupIcon.vue'
</script>

<template>
  <GroupIcon></GroupIcon>
</template>
