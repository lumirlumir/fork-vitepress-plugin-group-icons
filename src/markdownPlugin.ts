import type MarkdownIt from 'markdown-it'
import type { StateCore } from 'markdown-it/index.js'

// eslint-disable-next-line ts/explicit-function-return-type
function groupIconRule(state: StateCore) {
  const groupNameRegex = /\[([^\]]+)\]/

  for (const token of state.tokens) {
    const match = token.info.match(groupNameRegex)
    if (match) {
      const iconName = match[1]
      token.info = token.info.replace(
        groupNameRegex,
          `[<GroupIcon name="${iconName}" />]`,
      )
    }
  }
}

// eslint-disable-next-line ts/explicit-function-return-type
export function groupIconPlugin(md: MarkdownIt) {
  md.core.ruler.push('vp-group-icons', groupIconRule)
}
