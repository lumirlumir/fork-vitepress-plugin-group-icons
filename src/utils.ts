export function isSetEqual(set1: Set<string>, set2: Set<string>) {
  try {
    if (set1.size !== set2.size) {
      return false
    }
    for (const item of set1) {
      if (!set2.has(item)) {
        return false
      }
    }
    return true
  }
  catch {
    return false
  }
}

export const namedIconMatchRegex = /\s~([^~]+)~/
