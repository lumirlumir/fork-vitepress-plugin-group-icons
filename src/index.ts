import type { App, Plugin } from 'vue'
import GroupIcon from './components/GroupIcon.vue'
import { groupIconPlugin } from './markdownPlugin'

const GroupIconComponent: Plugin = {
  install: (app: App, options?: { foo: string }) => {
    app.config.globalProperties.$foo = options?.foo
    app.component('GroupIcon', GroupIcon)
  },
}

export {
  GroupIconComponent,
  groupIconPlugin,
}
