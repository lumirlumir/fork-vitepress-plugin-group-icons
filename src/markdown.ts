import type Markdown from 'markdown-it'
import { namedIconMatchRegex } from './utils'

interface MdPluginOptions {
  titleBar: {
    /**
     * Whether the title bar is included in the [Snippets](https://vitepress.dev/guide/markdown#import-code-snippets)
     *
     * @defaultValue false
     */
    includeSnippet?: boolean
  }
}

export function groupIconMdPlugin(md: Markdown, options?: MdPluginOptions) {
  const _options = options || { titleBar: { includeSnippet: false } }

  // code group rule
  const labelRE = /<label\b(?![^>]+\bdata-title\b)[^>]*>(.*?)<\/label>/g
  const codeGroupOpenRule = md.renderer.rules['container_code-group_open']
  if (codeGroupOpenRule) {
    md.renderer.rules['container_code-group_open'] = (...args) => {
      return codeGroupOpenRule(...args).replace(
        labelRE,
        (match: string, label: string) =>
          `<label data-title="${md.utils.escapeHtml(label)}"${match.slice(6)}`,
      )
    }
  }

  // replace named icon in label content
  const namedIconLabelMatchRegex = /(<label[^>]*>)(.*?)(~[^~]+~)(.*?)(<\/label>)/g
  md.renderer.rules['container_code-group_open']! = (...args) => {
    const code = codeGroupOpenRule!(...args)
    return code.replace(
      namedIconLabelMatchRegex,
      (_, labelStart, beforeIcon, _icon, afterIcon, labelEnd) => {
        return `${labelStart}${beforeIcon.trim()}${afterIcon}${labelEnd}`
      },
    )
  }

  // code block rule
  const fenceRule = md.renderer.rules.fence
  if (fenceRule) {
    md.renderer.rules.fence = (...args) => {
      const [tokens, idx] = args
      const token = tokens[idx]

      let isOnCodeGroup = false
      for (let i = idx - 1; i >= 0; i--) {
        if (tokens[i].type === 'container_code-group_open') {
          isOnCodeGroup = true
          break
        }

        if (tokens[i].type === 'container_code-group_close') {
          break
        }
      }
      const title = token.info.match(/\[(.*?)\]/)

      const isIncludedSnippet = _options.titleBar.includeSnippet

      // only render code block not in code-group
      if (!isOnCodeGroup && title && (!(token as any).src || isIncludedSnippet)) {
        const namedIconMatch = title[1].match(namedIconMatchRegex)
        return `<div class="vp-code-block-title">
      <div class="vp-code-block-title-bar">
          <span class="vp-code-block-title-text" data-title="${md.utils.escapeHtml(title[1])}">${namedIconMatch ? title[1].replace(namedIconMatch[0], '') : title[1]}</span>
      </div>
        ${fenceRule(...args)}
      </div>
      `
      }

      return fenceRule(...args)
    }
  }
}
