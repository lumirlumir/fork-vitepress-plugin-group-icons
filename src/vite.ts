import type { Plugin, ViteDevServer } from 'vite'
import { createFilter } from 'vite'
import { generateCSS } from './codegen'
import { isSetEqual } from './utils'

export interface Options {
  customIcon: Record<string, string>
  defaultLabels?: string[]
}

const filter = createFilter(
  [/\.md$/, /\.md\?vue/, /\.md\?v=/],
)

export function groupIconVitePlugin(options?: Options): Plugin {
  const virtualCssId = 'virtual:group-icons.css'
  const resolvedVirtualCssId = `\0${virtualCssId}`
  const combinedRegex = /\bdata-title=\\"([^"]*)\\"|\bdata-title="([^"]*)"|"data-title":\s*"([^"]*)"/g
  const matches = new Set<string>()

  let oldMatches: Set<string>
  let server: ViteDevServer | undefined

  options = options || { customIcon: {} }

  function handleUpdateModule() {
    const mod = server?.moduleGraph.getModuleById(resolvedVirtualCssId)
    if (mod) {
      server!.moduleGraph.invalidateModule(mod)
      server!.reloadModule(mod)
    }
  }

  return {
    name: 'vitepress-plugin-group-icons',
    enforce: 'post',
    resolveId(id) {
      if (id === virtualCssId) {
        return resolvedVirtualCssId
      }
      return undefined
    },

    configureServer(_server) {
      server = _server
    },

    async load(id) {
      if (id === resolvedVirtualCssId) {
        const { css } = await generateCSS(matches, options)
        oldMatches = new Set(matches)
        return css
      }

      return undefined
    },
    transform(code, id) {
      if (!filter(id))
        return

      while (true) {
        const match = combinedRegex.exec(code)
        if (!match)
          break

        matches.add(match[1] || match[2] || match[3])
      }

      if (!isSetEqual(matches, oldMatches)) {
        handleUpdateModule()
      }
    },
  }
}
