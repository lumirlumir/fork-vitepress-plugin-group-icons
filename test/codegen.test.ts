import { localIconLoader } from 'src'
import { generateCSS } from 'src/codegen'
import { describe, expect, it } from 'vitest'

describe('generate css', () => {
  it('builtin icon', async () => {
    const labels = new Set(['npm', 'yarn'])
    expect(await generateCSS(labels, {
      customIcon: {},
    })).toMatchSnapshot()
  })

  it('custom icon', { timeout: 0 }, async () => {
    const labels = new Set(['vitepress', 'ae', 'oxc'])
    expect(await generateCSS(labels, {
      customIcon: {
        ae: 'logos:adobe-after-effects',
        vitepress: localIconLoader(import.meta.url, '../docs/assets/vitepress.svg'),
      },
    })).toMatchSnapshot()
  })

  it('duplicate label', async () => {
    const labels = new Set(['foo.ts', 'bar.ts'])
    expect(await generateCSS(labels, {
      customIcon: {},
    })).toMatchSnapshot()
  })

  it('default labels', async () => {
    const labels = new Set(['npm'])
    expect(await generateCSS(labels, {
      customIcon: {},
      defaultLabels: ['yarn'],
    })).toMatchSnapshot()
  })

  it('default labels with custom icon', async () => {
    const labels = new Set(['npm'])
    expect(await generateCSS(labels, {
      customIcon: {
        curl: 'logos:curl',
      },
      defaultLabels: ['curl'],
    })).toMatchSnapshot()
  })
})
