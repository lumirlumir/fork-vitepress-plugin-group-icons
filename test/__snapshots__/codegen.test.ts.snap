// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate css > builtin icon 1`] = `
{
  "css": "
.vp-code-block-title [data-title]::before,
.vp-code-group [data-title]::before {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
  margin-bottom: -0.2em;
  background: var(--icon) no-repeat center / contain;
}

.vp-code-block-title-bar {
  position: relative;
  margin: 16px -24px 0 -24px;
  background-color: var(--vp-code-block-bg);
  overflow-x: auto;
  font-size: 14px;
  font-weight: 500;
  color: var(--vp-code-tab-text-color);
  white-space: nowrap;
  transition: background-color 0.5s;
  border-radius: 8px 8px 0 0;
  padding:0 12px;
  box-shadow: inset 0 -1px var(--vp-code-tab-divider);
}

.custom-block .vp-code-block-title-bar {
  margin: 16px 0 0 0;
}

@media (min-width: 640px) {
  .vp-code-block-title-bar {
    margin: 16px 0 0 0;
  }
}

.vp-code-block-title-text {
  padding: 0 12px;
  line-height: 48px;
}


.vp-code-block-title div[class*=language-] {
  margin-top: 0 !important;
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}

[data-title='npm']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%23c12127' d='M2 2h28v28H2'/%3E%3Cpath fill='%23fff' d='M7.25 7.25h17.5v17.5h-3.5v-14H16v14H7.25'/%3E%3C/svg%3E");
}
[data-title='yarn']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%232188b6' d='M28.208 24.409a10.5 10.5 0 0 0-3.959 1.822a23.7 23.7 0 0 1-5.835 2.642a1.63 1.63 0 0 1-.983.55a62 62 0 0 1-6.447.577c-1.163.009-1.876-.3-2.074-.776a1.573 1.573 0 0 1 .866-2.074a4 4 0 0 1-.514-.379c-.171-.171-.352-.514-.406-.388c-.225.55-.343 1.894-.947 2.5c-.83.839-2.4.559-3.328.072c-1.019-.541.072-1.813.072-1.813a.73.73 0 0 1-.992-.343a4.85 4.85 0 0 1-.667-2.949a5.37 5.37 0 0 1 1.749-2.895a9.3 9.3 0 0 1 .658-4.4a10.45 10.45 0 0 1 3.165-3.661S6.628 10.747 7.35 8.817c.469-1.262.658-1.253.812-1.308a3.6 3.6 0 0 0 1.452-.857a5.27 5.27 0 0 1 4.41-1.7S15.2 1.4 16.277 2.09a18.4 18.4 0 0 1 1.533 2.886s1.281-.748 1.425-.469a11.33 11.33 0 0 1 .523 6.132a14 14 0 0 1-2.6 5.411c-.135.225 1.551.938 2.615 3.887c.983 2.7.108 4.96.262 5.212c.027.045.036.063.036.063s1.127.09 3.391-1.308a8.5 8.5 0 0 1 4.277-1.604a1.081 1.081 0 0 1 .469 2.11Z'/%3E%3C/svg%3E");
}",
}
`;

exports[`generate css > custom icon 1`] = `
{
  "css": "
.vp-code-block-title [data-title]::before,
.vp-code-group [data-title]::before {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
  margin-bottom: -0.2em;
  background: var(--icon) no-repeat center / contain;
}

.vp-code-block-title-bar {
  position: relative;
  margin: 16px -24px 0 -24px;
  background-color: var(--vp-code-block-bg);
  overflow-x: auto;
  font-size: 14px;
  font-weight: 500;
  color: var(--vp-code-tab-text-color);
  white-space: nowrap;
  transition: background-color 0.5s;
  border-radius: 8px 8px 0 0;
  padding:0 12px;
  box-shadow: inset 0 -1px var(--vp-code-tab-divider);
}

.custom-block .vp-code-block-title-bar {
  margin: 16px 0 0 0;
}

@media (min-width: 640px) {
  .vp-code-block-title-bar {
    margin: 16px 0 0 0;
  }
}

.vp-code-block-title-text {
  padding: 0 12px;
  line-height: 48px;
}


.vp-code-block-title div[class*=language-] {
  margin-top: 0 !important;
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}

[data-title='ae']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1.03em' height='1em' viewBox='0 0 256 250'%3E%3Crect width='256' height='249.6' fill='%2300005B' rx='42.5'/%3E%3Cpath fill='%2399F' d='M102.843 149.333H63.172L55.1 174.454a2.02 2.02 0 0 1-1.901 1.547L32.947 176q-1.72 0-1.202-1.89l34.347-98.509l.343-1.035l.114-.354l.23-.74c.114-.382.228-.787.343-1.226c.45-2.291.68-4.62.687-6.955a1.063 1.063 0 0 1 1.202-1.202h27.306q1.2 0 1.374.859l38.983 109.335q.515 1.718-1.03 1.717h-22.326a1.59 1.59 0 0 1-1.717-1.202zm-33.488-21.181h27.134l-.387-1.277l-.621-2.022l-.909-2.896l-1.745-5.467l-1.662-5.254l-3.092-9.789a336 336 0 0 1-2.46-8.138l-.878-3.096l-.948-3.393l-.866-3.147h-.172a139 139 0 0 1-3.136 12.23l-1.16 3.723l-2.79 9l-1.414 4.585q-.383 1.242-.765 2.456l-.76 2.398l-.756 2.342l-.752 2.284l-.748 2.227q-.559 1.65-1.113 3.234m130.518 11.404h-33.831a24.1 24.1 0 0 0 3.263 9.461a17.67 17.67 0 0 0 7.813 6.44a32.6 32.6 0 0 0 13.653 2.676a62 62 0 0 0 11.077-1.216a40.8 40.8 0 0 0 9.523-2.482q.859-.685.86.858v16.315c.028.445-.061.889-.259 1.288c-.196.312-.46.576-.772.773a43.7 43.7 0 0 1-10.64 3.17a75.5 75.5 0 0 1-15.113 1.287q-11.814 0-19.917-3.545l-.52-.233a36.4 36.4 0 0 1-13.394-10.132a39.7 39.7 0 0 1-7.385-13.996A55.1 55.1 0 0 1 142 134.678a54.1 54.1 0 0 1 2.662-16.916a44.2 44.2 0 0 1 7.985-14.597a38.4 38.4 0 0 1 12.88-10.133c5.037-2.46 10.991-3.343 17.86-3.343a38.8 38.8 0 0 1 16.573 3.263a29.2 29.2 0 0 1 11.248 8.495a39 39 0 0 1 6.354 12.107a43.2 43.2 0 0 1 2.061 13.052q0 3.78-.257 6.87l-.153 1.764l-.16 1.674l-.095.848l-.022.179a1.556 1.556 0 0 1-1.545 1.374l-.471.01l-.587.03l-.457.033l-1.068.091l-.938.093q-.383.04-.795.075l-.851.067l-.909.06l-.965.05l-1.022.043l-.533.018l-1.107.03c-1.095.027-2.217-.041-3.368-.125l-1.746-.128a40 40 0 0 0-2.7-.106m-33.831-15.645h23.386l1.992-.013l1.354-.019l.877-.02l.984-.034a6 6 0 0 0 2.49-.825v-1.03a13.7 13.7 0 0 0-.686-3.95a14.03 14.03 0 0 0-13.91-9.79a14.92 14.92 0 0 0-14.169 8.072a24.8 24.8 0 0 0-2.318 7.61'/%3E%3C/svg%3E");
}
[data-title='vitepress']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='48' height='48' viewBox='0 0 48 48' fill='none'%3E%3Clink xmlns=''/%3E%3Cscript xmlns=''/%3E %3Cpath d='M5.03628 7.87818C4.75336 5.83955 6.15592 3.95466 8.16899 3.66815L33.6838 0.0367403C35.6969 -0.24977 37.5581 1.1706 37.841 3.20923L42.9637 40.1218C43.2466 42.1604 41.8441 44.0453 39.831 44.3319L14.3162 47.9633C12.3031 48.2498 10.4419 46.8294 10.159 44.7908L5.03628 7.87818Z' fill='url(%23paint0_linear_1287_1214)'/%3E %3Cpath d='M6.85877 7.6188C6.71731 6.59948 7.41859 5.65703 8.42512 5.51378L33.9399 1.88237C34.9465 1.73911 35.8771 2.4493 36.0186 3.46861L41.1412 40.3812C41.2827 41.4005 40.5814 42.343 39.5749 42.4862L14.0601 46.1176C13.0535 46.2609 12.1229 45.5507 11.9814 44.5314L6.85877 7.6188Z' fill='white'/%3E %3Cpath d='M33.1857 14.9195L25.8505 34.1576C25.6991 34.5547 25.1763 34.63 24.9177 34.2919L12.3343 17.8339C12.0526 17.4655 12.3217 16.9339 12.7806 16.9524L22.9053 17.3607C22.9698 17.3633 23.0344 17.3541 23.0956 17.3337L32.5088 14.1992C32.9431 14.0546 33.3503 14.4878 33.1857 14.9195Z' fill='url(%23paint1_linear_1287_1214)'/%3E %3Cpath d='M27.0251 12.5756L19.9352 15.0427C19.8187 15.0832 19.7444 15.1986 19.7546 15.3231L20.3916 23.063C20.4066 23.2453 20.5904 23.3628 20.7588 23.2977L22.7226 22.5392C22.9064 22.4682 23.1021 22.6138 23.0905 22.8128L22.9102 25.8903C22.8982 26.0974 23.1093 26.2436 23.295 26.1567L24.4948 25.5953C24.6808 25.5084 24.892 25.6549 24.8795 25.8624L24.5855 30.6979C24.5671 31.0004 24.9759 31.1067 25.1013 30.8321L25.185 30.6487L29.4298 17.8014C29.5008 17.5863 29.2968 17.3809 29.0847 17.454L27.0519 18.1547C26.8609 18.2205 26.6675 18.0586 26.6954 17.8561L27.3823 12.8739C27.4103 12.6712 27.2163 12.5091 27.0251 12.5756Z' fill='url(%23paint2_linear_1287_1214)'/%3E %3Cdefs%3E %3ClinearGradient id='paint0_linear_1287_1214' x1='6.48163' y1='1.9759' x2='39.05' y2='48.2064' gradientUnits='userSpaceOnUse'%3E %3Cstop stop-color='%2349C7FF'/%3E %3Cstop offset='1' stop-color='%23BD36FF'/%3E %3C/linearGradient%3E %3ClinearGradient id='paint1_linear_1287_1214' x1='11.8848' y1='16.4266' x2='26.7246' y2='31.4177' gradientUnits='userSpaceOnUse'%3E %3Cstop stop-color='%2341D1FF'/%3E %3Cstop offset='1' stop-color='%23BD34FE'/%3E %3C/linearGradient%3E %3ClinearGradient id='paint2_linear_1287_1214' x1='21.8138' y1='13.7046' x2='26.2464' y2='28.8069' gradientUnits='userSpaceOnUse'%3E %3Cstop stop-color='%23FFEA83'/%3E %3Cstop offset='0.0833333' stop-color='%23FFDD35'/%3E %3Cstop offset='1' stop-color='%23FFA800'/%3E %3C/linearGradient%3E %3C/defs%3E %3Cscript xmlns=''/%3E%3C/svg%3E ");
}",
}
`;

exports[`generate css > default labels 1`] = `
{
  "css": "
.vp-code-block-title [data-title]::before,
.vp-code-group [data-title]::before {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
  margin-bottom: -0.2em;
  background: var(--icon) no-repeat center / contain;
}

.vp-code-block-title-bar {
  position: relative;
  margin: 16px -24px 0 -24px;
  background-color: var(--vp-code-block-bg);
  overflow-x: auto;
  font-size: 14px;
  font-weight: 500;
  color: var(--vp-code-tab-text-color);
  white-space: nowrap;
  transition: background-color 0.5s;
  border-radius: 8px 8px 0 0;
  padding:0 12px;
  box-shadow: inset 0 -1px var(--vp-code-tab-divider);
}

.custom-block .vp-code-block-title-bar {
  margin: 16px 0 0 0;
}

@media (min-width: 640px) {
  .vp-code-block-title-bar {
    margin: 16px 0 0 0;
  }
}

.vp-code-block-title-text {
  padding: 0 12px;
  line-height: 48px;
}


.vp-code-block-title div[class*=language-] {
  margin-top: 0 !important;
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}

[data-title='npm']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%23c12127' d='M2 2h28v28H2'/%3E%3Cpath fill='%23fff' d='M7.25 7.25h17.5v17.5h-3.5v-14H16v14H7.25'/%3E%3C/svg%3E");
}
[data-title='yarn']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%232188b6' d='M28.208 24.409a10.5 10.5 0 0 0-3.959 1.822a23.7 23.7 0 0 1-5.835 2.642a1.63 1.63 0 0 1-.983.55a62 62 0 0 1-6.447.577c-1.163.009-1.876-.3-2.074-.776a1.573 1.573 0 0 1 .866-2.074a4 4 0 0 1-.514-.379c-.171-.171-.352-.514-.406-.388c-.225.55-.343 1.894-.947 2.5c-.83.839-2.4.559-3.328.072c-1.019-.541.072-1.813.072-1.813a.73.73 0 0 1-.992-.343a4.85 4.85 0 0 1-.667-2.949a5.37 5.37 0 0 1 1.749-2.895a9.3 9.3 0 0 1 .658-4.4a10.45 10.45 0 0 1 3.165-3.661S6.628 10.747 7.35 8.817c.469-1.262.658-1.253.812-1.308a3.6 3.6 0 0 0 1.452-.857a5.27 5.27 0 0 1 4.41-1.7S15.2 1.4 16.277 2.09a18.4 18.4 0 0 1 1.533 2.886s1.281-.748 1.425-.469a11.33 11.33 0 0 1 .523 6.132a14 14 0 0 1-2.6 5.411c-.135.225 1.551.938 2.615 3.887c.983 2.7.108 4.96.262 5.212c.027.045.036.063.036.063s1.127.09 3.391-1.308a8.5 8.5 0 0 1 4.277-1.604a1.081 1.081 0 0 1 .469 2.11Z'/%3E%3C/svg%3E");
}",
}
`;

exports[`generate css > default labels with custom icon 1`] = `
{
  "css": "
.vp-code-block-title [data-title]::before,
.vp-code-group [data-title]::before {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
  margin-bottom: -0.2em;
  background: var(--icon) no-repeat center / contain;
}

.vp-code-block-title-bar {
  position: relative;
  margin: 16px -24px 0 -24px;
  background-color: var(--vp-code-block-bg);
  overflow-x: auto;
  font-size: 14px;
  font-weight: 500;
  color: var(--vp-code-tab-text-color);
  white-space: nowrap;
  transition: background-color 0.5s;
  border-radius: 8px 8px 0 0;
  padding:0 12px;
  box-shadow: inset 0 -1px var(--vp-code-tab-divider);
}

.custom-block .vp-code-block-title-bar {
  margin: 16px 0 0 0;
}

@media (min-width: 640px) {
  .vp-code-block-title-bar {
    margin: 16px 0 0 0;
  }
}

.vp-code-block-title-text {
  padding: 0 12px;
  line-height: 48px;
}


.vp-code-block-title div[class*=language-] {
  margin-top: 0 !important;
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}

[data-title='curl']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='3.33em' height='1em' viewBox='0 0 512 154'%3E%3Cpath fill='%230C544C' d='M498.445 20.959a7.403 7.403 0 0 1-7.404-7.404a7.4 7.4 0 0 1 7.404-7.404a7.403 7.403 0 1 1 0 14.808m-73.532 126.503a7.403 7.403 0 0 1-7.404-7.404a7.4 7.4 0 0 1 7.404-7.404a7.4 7.4 0 0 1 7.398 7.404c0 4.086-3.311 7.404-7.398 7.404M498.445 0c-7.485 0-13.554 6.069-13.554 13.555c0 1.597.402 3.084.91 4.506l-63.587 108.984c-6.145 1.289-10.856 6.483-10.856 13.013c0 7.485 6.07 13.554 13.555 13.554c7.48 0 13.549-6.069 13.549-13.554c0-1.504-.403-2.88-.852-4.233l63.908-109.34C507.476 25.057 512 19.95 512 13.555C512 6.069 505.931 0 498.445 0'/%3E%3Cpath fill='%23073551' d='M442.986 20.959a7.403 7.403 0 0 1-7.404-7.404a7.4 7.4 0 0 1 7.404-7.404a7.4 7.4 0 0 1 7.398 7.404c0 4.086-3.312 7.404-7.398 7.404m-73.539 126.503a7.403 7.403 0 1 1 0-14.808a7.4 7.4 0 0 1 7.404 7.404a7.403 7.403 0 0 1-7.404 7.404M442.986 0c-7.492 0-13.555 6.069-13.555 13.555c0 1.597.402 3.084.91 4.506l-63.587 108.984c-6.15 1.289-10.861 6.483-10.861 13.013c0 7.485 6.068 13.554 13.554 13.554s13.555-6.069 13.555-13.554c0-1.504-.403-2.88-.852-4.233l63.908-109.34c5.958-1.428 10.482-6.535 10.482-12.93C456.54 6.069 450.465 0 442.986 0M341.708 84.831c5.988 0 10.902 3.941 12.698 9.322c.449 1.353.857 2.728.857 4.233c0 7.48-6.069 13.554-13.555 13.554s-13.554-6.075-13.554-13.554c0-6.53 4.71-11.724 10.855-13.013c.904-.186 1.743-.542 2.7-.542m0 6.15a7.403 7.403 0 0 0-7.404 7.405c0 4.086 3.318 7.398 7.404 7.398s7.404-3.312 7.404-7.398a7.403 7.403 0 0 0-7.404-7.404m0-53.512c5.982 0 10.902 3.94 12.698 9.322c.449 1.352.857 2.734.857 4.232c0 7.486-6.069 13.555-13.555 13.555s-13.554-6.069-13.554-13.555c0-6.53 4.71-11.724 10.855-13.006c.904-.193 1.743-.548 2.7-.548m0 6.15a7.403 7.403 0 0 0-7.404 7.404a7.403 7.403 0 1 0 14.808 0a7.407 7.407 0 0 0-7.404-7.404M72.754 56.877c5.662.038 11.593 1.367 11.593 8.605c0 7.532-6.81 8.628-11.846 8.628H42.185c-3.428 0-6.832 1.795-9.887 4.29l-10.4 8.792c-2.706 2.169-4.665 3.964-4.665 7.37v21.36c0 3.405 1.96 5.2 4.664 7.346l10.4 8.791c3.056 2.519 6.46 4.314 9.888 4.314h30.316c5.037 0 11.846 1.073 11.846 8.629c0 7.345-6.11 8.604-11.846 8.604H44.262c-9.366-.018-13.484-.4-21.292-6.809l-12.733-10.61C1.982 129.377 0 124.177 0 113.59V96.893c0-10.587 1.982-15.787 10.237-22.62L22.97 63.686c7.808-6.41 11.926-6.79 21.292-6.809Zm96.313 0c7.346 0 8.605 5.923 8.605 11.846v72.292c.002.43.01.91.013 1.43v.529q-.001.36-.01.737l-.013.38c-.16 4.215-1.315 9.515-8.431 9.515c-7.346 0-8.605-6.11-8.605-12.032v-2.869l-13.992 9.165c-5.574 3.591-10.237 5.736-14.901 5.736H119c-4.85 0-10.237-1.445-15.974-6.273c-7.719-6.646-7.555-14.015-7.905-23.342l-1.796-54.195c-.186-6.273.723-12.92 8.605-12.92c7.37 0 8.255 6.274 8.442 11.847l1.795 52.936c.187 3.94.35 8.628 1.796 11.123c1.446 2.519 5.923 3.591 10.05 3.591h5.737q2.974 0 5.387-1.609l25.302-15.974V68.462c.038-5.835 1.368-11.585 8.628-11.585m77.352 0c6.646 0 12.196 2.145 16.86 6.809c5.736 5.76 7.718 11.683 7.718 19.565c0 5.923-1.072 12.033-8.441 12.033c-6.273 0-8.792-5.037-8.792-10.587c0-2.868 0-5.037-1.795-7.532c-2.036-2.899-5.855-3.047-9.766-3.055h-5.672c-3.218 0-5.55 1.609-8.605 4.127l-24.042 19.915v43.608c0 5.923-1.26 11.846-8.629 11.846c-7.345 0-8.605-5.923-8.605-11.846V69.355c.021-6.192.83-12.478 8.605-12.478c6.833 0 8.629 5.736 8.629 11.473v7.02l13.455-11.148c4.85-3.94 9.864-7.345 15.251-7.345Zm43.758-35.539c9.598.04 12.402 1.91 12.402 12.032v103.003h12.92c5.76 0 11.846 1.26 11.846 8.629c0 7.532-6.81 8.604-11.846 8.604h-43.991c-5.013-.032-11.627-1.181-11.627-8.604c0-7.37 6.087-8.629 11.847-8.629h13.642V38.571h-13.642c-5.037 0-11.847-1.073-11.847-8.629c0-7.345 6.087-8.604 11.847-8.604Z'/%3E%3C/svg%3E");
}
[data-title='npm']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%23c12127' d='M2 2h28v28H2'/%3E%3Cpath fill='%23fff' d='M7.25 7.25h17.5v17.5h-3.5v-14H16v14H7.25'/%3E%3C/svg%3E");
}",
}
`;

exports[`generate css > duplicate label 1`] = `
{
  "css": "
.vp-code-block-title [data-title]::before,
.vp-code-group [data-title]::before {
  display: inline-block;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
  margin-bottom: -0.2em;
  background: var(--icon) no-repeat center / contain;
}

.vp-code-block-title-bar {
  position: relative;
  margin: 16px -24px 0 -24px;
  background-color: var(--vp-code-block-bg);
  overflow-x: auto;
  font-size: 14px;
  font-weight: 500;
  color: var(--vp-code-tab-text-color);
  white-space: nowrap;
  transition: background-color 0.5s;
  border-radius: 8px 8px 0 0;
  padding:0 12px;
  box-shadow: inset 0 -1px var(--vp-code-tab-divider);
}

.custom-block .vp-code-block-title-bar {
  margin: 16px 0 0 0;
}

@media (min-width: 640px) {
  .vp-code-block-title-bar {
    margin: 16px 0 0 0;
  }
}

.vp-code-block-title-text {
  padding: 0 12px;
  line-height: 48px;
}


.vp-code-block-title div[class*=language-] {
  margin-top: 0 !important;
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}

[data-title='foo.ts']::before,[data-title='bar.ts']::before {
  content: '';
  --icon: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='1em' height='1em' viewBox='0 0 32 32'%3E%3Cpath fill='%23007acc' d='M23.827 8.243a4.4 4.4 0 0 1 2.223 1.281a6 6 0 0 1 .852 1.143c.011.045-1.534 1.083-2.471 1.662c-.034.023-.169-.124-.322-.35a2.01 2.01 0 0 0-1.67-1c-1.077-.074-1.771.49-1.766 1.433a1.3 1.3 0 0 0 .153.666c.237.49.677.784 2.059 1.383c2.544 1.095 3.636 1.817 4.31 2.843a5.16 5.16 0 0 1 .416 4.333a4.76 4.76 0 0 1-3.932 2.815a11 11 0 0 1-2.708-.028a6.53 6.53 0 0 1-3.616-1.884a6.3 6.3 0 0 1-.926-1.371a3 3 0 0 1 .327-.208c.158-.09.756-.434 1.32-.761l1.024-.6l.214.312a4.8 4.8 0 0 0 1.35 1.292a3.3 3.3 0 0 0 3.458-.175a1.545 1.545 0 0 0 .2-1.974c-.276-.395-.84-.727-2.443-1.422a8.8 8.8 0 0 1-3.349-2.055a4.7 4.7 0 0 1-.976-1.777a7.1 7.1 0 0 1-.062-2.268a4.33 4.33 0 0 1 3.644-3.374a9 9 0 0 1 2.691.084m-8.343 1.483l.011 1.454h-4.63v13.148H7.6V11.183H2.97V9.755a14 14 0 0 1 .04-1.466c.017-.023 2.832-.034 6.245-.028l6.211.017Z'/%3E%3C/svg%3E");
}",
}
`;
